DataFormats/TestObjects/BuildFile.xml

```diff
@@ -2,6 +2,15 @@
 <use name="DataFormats/Provenance"/>
 <use name="FWCore/SOA" source_only="1"/>
 <use name="FWCore/Utilities"/>
+<flags LCG_DICT_HEADER="classes.h"/>
+<flags LCG_DICT_XML="classes_def.xml"/>
+
+<iftool name="cuda-gcc-support">
+  <use name="cuda"/>
+  <flags LCG_DICT_HEADER="classes_cuda.h"/>
+  <flags LCG_DICT_XML="classes_cuda_def.xml"/>
+</iftool>
+
```

Comment:
It seems to me the test (which itself is great!) doesn't really need the CUDA dependence, and the necessary and sufficient component of the test is the omission of the dictionary for `edm::Wrapper<edmtest::MissingDictionaryCUDAObject>`. Without CUDA dependence, and with the auto-parsing enabled, the test job would then succeed, after which the test itself would fail.

If I understood correctly, could you remove the dependence on CUDA (including names) from `DataFormats/TestObjects` and `FWCore/Integration`?

If you want to have a CUDA-specific test, it could be placed e.g. in `HeterogeneousCore/CUDATest`.

Reply:
If I comment out
```c++
//#include <cuda_runtime.h>
```
in `DataFormats/TestObjects/src/classes_cuda.h` and run the test _without the changes_ to disable the ROOT autoloading, I already get the "nice" message:
```
$ cmsRun FWCore/Integration/test/testMissingDictionaryCUDAChecking_cfg.py 
----- Begin Fatal Exception 18-Dec-2024 21:10:30 CET-----------------------
An exception of category 'DictionaryNotFound' occurred while
   [0] Constructing the EventProcessor
   [1] Calling ProductRegistry::initializeLookupTables, checking dictionaries for consumed products
Exception Message:
No data dictionary found for the following classes:

  edm::Wrapper<edmtest::MissingDictionaryCUDAObject>

Most likely each dictionary was never generated, but it may
be that it was generated in the wrong package. Please add
(or move) the specification '<class name="whatever"/>' to
the appropriate classes_def.xml file along with any other
information needed there. For example, if this class has any
transient members, you need to specify them in classes_def.xml.
Also include the class header in classes.h

The list of types above was generated while checking for
dictionaries related to products declared to be consumed.
A type listed above might or might not be a type declared
to be consumed. Instead it might be the type of a data member,
base class, wrapped type or other object needed by a consumed
type.  Below is some additional information which lists
the types declared to be consumed by a module and which
are associated with the types whose dictionaries were not
found:

  edmtest::MissingDictionaryCUDAObject

----- End Fatal Exception -------------------------------------------------
```

Which is the behaviour that we already get for missing dictionaries, normally.
So, without the CUDA include, the new test already succeeds without this PR.

It is the CUDA include (potentially other externals, too - I used CUDA because this is how we noticed the issue in the first place) that makes ROOT fail with a different exception
```
DataFormatsTestObjects_x1r dictionary payload:59:10: remark: could not acquire lock file for module 'cuda': failed to create unique file /data/cmssw/el8_amd64_gcc12/lcg/root/6.30.09-0fd387670244508fd2077d7feaf8b1c8/lib/cuda.pcm.lock-1d54d871: Permission denied [-Rmodule-build]
#include <cuda_runtime.h>
         ^
DataFormatsTestObjects_x1r dictionary payload:59:10: remark: building module 'cuda' as '/data/cmssw/el8_amd64_gcc12/lcg/root/6.30.09-0fd387670244508fd2077d7feaf8b1c8/lib/cuda.pcm' [-Rmodule-build]
error: unable to open output file '/data/cmssw/el8_amd64_gcc12/lcg/root/6.30.09-0fd387670244508fd2077d7feaf8b1c8/lib/cuda.pcm': 'Permission denied'
DataFormatsTestObjects_x1r dictionary payload:59:10: remark: finished building module 'cuda' [-Rmodule-build]
DataFormatsTestObjects_x1r dictionary payload:59:10: fatal error: could not build module 'cuda'
#include <cuda_runtime.h>
 ~~~~~~~~^
----- Begin Fatal Exception 18-Dec-2024 21:15:26 CET-----------------------
An exception of category 'FatalRootError' occurred while
   [0] Constructing the EventProcessor
   [1] Calling edm::TypeWithDict::byName()
   Additional Info:
      [a] Getting TClass for edm::Wrapper<edmtest::MissingDictionaryCUDAObject>
      [b] Fatal Root Error: @SUB=TInterpreter::AutoParse
Error parsing payload code for class edmtest::MissingDictionaryCUDAObject with content:
...
```
which prevents the framework from showing the "nice" message.

Reply:
In fact, I tried with a few other externals (zlib, boost, tbb) and none of them caused the weird exception that we get from CUDA.

Reply:
> If you want to have a CUDA-specific test, it could be placed e.g. in `HeterogeneousCore/CUDATest`.

Sure, I can move the test if you prefer.

What about the test data formats ?

Reply:
> in `DataFormats/TestObjects/src/classes_cuda.h` and run the test _without the changes_ to disable the ROOT autoloading, I already get the "nice" message:

Darn :( I'm confused why that happens though. I _guess_ in that case ROOT succeeds in the auto parsing itself, but the auto parsing does not lead to a dictionary for `edm::Wrapper<...>` (according to https://root.cern.ch/doc/master/classTClass.html information "From a lookup in the AST built by cling" "does not carry the information which is necessary to ROOT to perform I/O of instances of the class/struct it represents", so maybe we catch that).

We have one existing test that fails if ROOT ends up auto parsing the headers [`IOPool/Input/test/testNoParentDictionary.sh`](https://github.com/cms-sw/cmssw/blob/master/IOPool/Input/test/testNoParentDictionary.sh), but the setup is quite heavy. If the pattern would be applied here, I'd expect something simpler to be sufficient. Maybe adjusting `ROOT_INCLUDE_PATH
https://github.com/cms-sw/cmssw/blob/666efe2ff1988126cebcb0a3aaad35b8d4d0312a/IOPool/Input/test/testNoParentDictionary.sh#L42
and making the relevant header to fail to compile
https://github.com/cms-sw/cmssw/blob/666efe2ff1988126cebcb0a3aaad35b8d4d0312a/IOPool/Input/test/testNoParentDictionary.sh#L49
would be enough.


> > If you want to have a CUDA-specific test, it could be placed e.g. in `HeterogeneousCore/CUDATest`.
> 
> Sure, I can move the test if you prefer.

Thank you.

> What about the test data formats ?

Given that they have a dependence on CUDA (even if only via dictionary compilation), they should be placed outside of `DataFormats/TestObjects` as well. I see we already have a transient test data format defined in `HeterogeneousCore/CUDATest`, so I'd think these test data formats could be moved there as well.